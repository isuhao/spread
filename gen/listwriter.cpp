#include "listwriter.hpp"
#include <mangle/stream/clients/copy_stream.hpp>
#include <stdexcept>
#include <assert.h>
#include <algorithm>
#include "unpack/dirwriter.hpp"
#include "hash/hash_stream.hpp"
#include "misc/readjson.hpp"

using namespace std;
using namespace Spread;
using namespace SpreadGen;
using namespace Mangle::VFS;
using namespace Mangle::Stream;

void ListWriter::write(const PackLister &lister, const string &where)
{
  StreamFactoryPtr ptr(new Unpack::DirWriter(where));
  write(lister, ptr);
}

/* Used to sort a list of strings before appending it to a Json
   array. The vector is sorted in place.
 */
void appendSorted(Json::Value &out, vector<string> &input)
{
  assert(out.isNull() || out.isArray());

  sort(input.begin(), input.end());

  int old = out.size();
  out.resize(old + input.size());
  for(int i=0; i<input.size(); i++)
    out[old+i] = input[i];
}

void ListWriter::write(const PackLister &lst, StreamFactoryPtr output)
{
  // Packs list
  {
    Json::Value packs;

    map<string, PackLister::PackInfo>::const_iterator it;
    for(it = lst.packs.begin(); it != lst.packs.end(); it++)
      {
        Json::Value dirs, total;

        const PackLister::PackInfo &pinf = it->second;
        assert(pinf.dirs.size() == pinf.paths.size());

        for(int i=0; i<pinf.dirs.size(); i++)
          {
            // Add the hash string first
            std::string out = pinf.dirs[i].toString();

            // Append the path string, if any
            std::string path = pinf.paths[i];
            if(path != "") out += " " + path;

            dirs.append(out);
          }

        if(dirs.size() == 0)
          throw runtime_error("Package '" + it->first + "' has no output directories.");
        total["dirs"] = dirs;

        if(pinf.version != "")
          total["version"] = pinf.version;

        packs[it->first] = total;
      }

    ReadJson::writeJson(output->open("packs.json"), packs);
  }

  // Rule file
  {
    Json::Value rules;

    {
      RuleList::const_iterator it;
      vector<string> tmp;
      tmp.reserve(lst.ruleSet.size());
      for(it = lst.ruleSet.begin(); it != lst.ruleSet.end(); it++)
        tmp.push_back((*it)->ruleString);

      /* Sort the rules before adding them. Otherwise they will be
         ordered by pointer order (since RuleList is a set<> of
         pointers.) This would make output non-deterministic, leading
         to unnecessary hash changes, uploads and refreshes.
       */
      appendSorted(rules, tmp);
    }
    {
      set<const ArcRuleData*>::const_iterator it;
      vector<string> tmp;
      tmp.reserve(lst.arcSet.size());
      for(it = lst.arcSet.begin(); it != lst.arcSet.end(); it++)
        tmp.push_back((*it)->ruleString);

      appendSorted(rules, tmp);
    }

    ReadJson::writeJson(output->open("rules.json"), rules);
  }

  /* 1k-cache. This is a lookup of hash snippets for the first 1 Kb of
     the data block for each hash. This is useful to quickly validate
     URL rules - it lets you determine if a remote file has changed by
     just downloading the first Kb, instead of the entire file.

     Since this is only intended to validate URL rules, we only
     include entries need for that.

     We don't really use these yet, so don't bother implementing them
     yet.

  {
    Json::Value out1k;

    ReadJson::writeJson(output->open("1kcache.json"), out1k);
  }
   */

  // Dir files
  {
    PackLister::HashSet::const_iterator it;
    for(it = lst.dirs.begin(); it != lst.dirs.end(); it++)
      {
        const Hash &dirHash = *it;
        const string &dirFile = cache.index.findHash(dirHash);

        if(dirFile == "")
          throw runtime_error("Couldn't find any source for directory object: " + dirHash.toString());

        HashStreamPtr inp(new HashStream(dirFile));
        StreamPtr out = output->open("dirs/" + dirHash.toString());
        assert(out);

        CopyStream::copy(inp, out);

        if(inp->finish() != dirHash)
          throw runtime_error("Hash mismatch in file " + dirFile + " (expected " + dirHash.toString() + ")");
      }
  }
}
