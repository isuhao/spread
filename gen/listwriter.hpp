#ifndef __SPREAD_GEN_LISTWRITER_HPP_
#define __SPREAD_GEN_LISTWRITER_HPP_

#include "packlister.hpp"
#include <mangle/vfs/stream_factory.hpp>
//#include "misc/jconfig.hpp"

/* The ListWriter takes the output from a PackLister and writes it as
   files to a directory.

   The current SR0-compatible format has the following files:

   - packs.json       index of packages
   - rules.json       rule file
   - 1kcache.json     list of 1k hashes (not implemented yet)
   - dirs/*           directory object files

   The format is optimized for readability and ease of use, not size.
   Future versions may be much more size-optimized. Specifically:

   - write all data in binary (optimized for small bsdiffs)
   - hashes written in one place, sorted and size-compressed in binary
     (sorting tends to increase compressibility)
   - all the rules, packs and dirs reference a shared hash list, but
     in a way so that references don't change needlessly between
     updates.
 */

namespace SpreadGen
{
  struct ListWriter
  {
    ListWriter(Cache::Cache &_cache/*, const Misc::JConfig &_c1k*/)
      : cache(_cache)/*, cache1k(_c1k)*/ {}

    // Write to a directory
    void write(const PackLister &lister, const std::string &where);

    // Write to a stream factory (or "virtual directory")
    void write(const PackLister &lister, Mangle::VFS::StreamFactoryPtr output);

  private:
    Cache::Cache &cache;
    //Misc::JConfig &cache1k;
  };
}

#endif
