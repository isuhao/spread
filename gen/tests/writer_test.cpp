#include "listwriter.hpp"
#include <iostream>
#include "cache/index.hpp"

#include "print_dir.hpp"

using namespace SpreadGen;
using namespace Spread;
using namespace std;

Cache::CacheIndex cache;
RuleSet rules;
PackLister lst(cache, rules);

Hash dirHash, dirHash2;
Hash hello("hello",5);
Hash world("world",5);

Hash arcHash("ARC_WITH_DIR");
Hash arcHash2("OTHER_ARC");

int main()
{
  Hash::DirMap dir;
  dir["world.txt"] = world;
  dirHash2 = Dir::write(dir, "_writer_dir2.out");

  dir["hello.txt"] = hello;
  dirHash = Dir::write(dir, "_writer_dir.out");

  cache.addFile("_writer_dir.out");
  cache.addFile("_writer_dir2.out");
  rules.addURL(hello, "url-to-hello");
  rules.addURL(hello, "other-hello-url", 3, 2);
  rules.addURL(Hash("blah"), "you will never see this");
  rules.addArchive(arcHash2, dirHash2);
  rules.addArchive(arcHash, dirHash);
  rules.addURL(arcHash2, "url-to-arc2", 12, 3.4);
  rules.addURL(arcHash, "url-to-arc");

  lst.addDir("no-hint", dirHash);
  lst.setVersion("no-hint", "1.2.3");
  //lst.addHint("with-hint", arcHash2);
  lst.addDir("with-hint", dirHash);

  ListWriter wrt(cache);

  wrt.write(lst, "_writer/");

  printDir("_writer");

  return 0;
}
