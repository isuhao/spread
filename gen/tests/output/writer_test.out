
Directory: _writer
00                                                 dirs
If6bDOAeOVoiKM9uWGt19noPySYSE-spItCkseTUX6U7       dirs/If6bDOAeOVoiKM9uWGt19noPySYSE-spItCkseTUX6U7
dD6saPLijaSBk4Q3b7XFg-n4h6DS7FAi7vdIg-b0Frxu       dirs/dD6saPLijaSBk4Q3b7XFg-n4h6DS7FAi7vdIg-b0Frxu
-3v9Xk6Vchyn2zyJk9Gy18Ut0xQNcrzRCpJN-Z1WyncDAQ     packs.json
B1aCrzu22Jl6TXCw2EqrpBelGw3B5xb2vjLLmXoCHc5hAQ     rules.json
Total 5 elements
Hash: QZ6Xa6KcWi6yiEviA-8naC5qcb7UxiMyIdX55vrIXiFUAQ
